query-engine/metrics/src/recorder.rs

```diff
@@ -1,122 +1,165 @@
-use std::sync::Arc;
+use std::sync::{Arc, OnceLock};
 
+use derive_more::Display;
 use metrics::{Counter, CounterFn, Gauge, GaugeFn, Histogram, HistogramFn, Key, Recorder, Unit};
 use metrics::{KeyName, Metadata, SharedString};
-use tracing::trace;
 
-use super::common::KeyLabels;
-use super::{METRIC_COUNTER, METRIC_DESCRIPTION, METRIC_GAUGE, METRIC_HISTOGRAM, METRIC_TARGET};
+use crate::common::{MetricAction, MetricType};
+use crate::registry::MetricVisitor;
+use crate::MetricRegistry;
 
-#[derive(Default)]
-pub(crate) struct MetricRecorder;
+static GLOBAL_RECORDER: OnceLock<Option<MetricRecorder>> = const { OnceLock::new() };
+
+#[derive(Display, Debug)]
+#[display(fmt = "global recorder can only be installed once")]
+pub struct AlreadyInstalled;
+
+impl std::error::Error for AlreadyInstalled {}
+
+fn set_global_recorder(recorder: MetricRecorder) -> Result<(), AlreadyInstalled> {
+    GLOBAL_RECORDER.set(Some(recorder)).map_err(|_| AlreadyInstalled)
+}
+
+pub(crate) fn global_recorder() -> Option<MetricRecorder> {
+    GLOBAL_RECORDER.get()?.clone()
+}
+
+#[derive(Clone)]
+pub struct MetricRecorder {
+    inner: Arc<Inner>,
+}
```

Comment:
It's funny how we discussed with Serhii on Friday that we don't like this pattern, and then I immediately had to use it on Monday because otherwise I would need `Arc<MetricRecorder>` to implement `Recorder` which isn't possible because `Arc` isn't a fundamental type like e.g. `Box`, so it's not allowed to implement a foreign trait even on your own `Arc`ed type.

Reply:
Actually this exact pattern isn't really necessary. Let's wrap the registry here directly. Technically we don't even need an Arc because `MetricsRegistry` is an `Arc` itself but it might be better to treat that as an implementation detail.